Client/game_sa/CModelInfoSA.cpp

```diff
@@ -1728,6 +1734,69 @@ void CModelInfoSA::MakeVehicleAutomobile(ushort usBaseID)
     CopyStreamingInfoFromModel(usBaseID);
 }
 
+bool GetFirstAtomic(RpAtomic* atomic, void* data)
+{
+    RpAtomic** firstAtomicPtr = (RpAtomic**)data;
+
+    if (*firstAtomicPtr == nullptr)
+        *firstAtomicPtr = atomic;
+
+    return false;
+}
+
+bool CModelInfoSA::MakeAtomicModel(ushort usModelId)
+{
+    if (GetModelType() != eModelInfoType::CLUMP)
+        return false;
+
+    CBaseModelInfoSAInterface*   pBaseObjectInfo = (CBaseModelInfoSAInterface*)ppModelInfo[usModelId];
+    CAtomicModelInfoSAInterface* m_pInterface = new CAtomicModelInfoSAInterface();
+    MemCpyFast(m_pInterface, pBaseObjectInfo, sizeof(CBaseModelInfoSAInterface));
+
+    CAtomicModelInfo_SA_VTBL* vfbl = (CAtomicModelInfo_SA_VTBL*)vftable_CAtomicModelModelInfo;
+    m_pInterface->VFTBL = vfbl;
+    RpClump*  pClump = (RpClump*)m_pInterface->pRwObject;
+    RpAtomic* firstAtomic = nullptr;
+    RpClumpForAllAtomics(pClump, GetFirstAtomic, &firstAtomic);
+    RwFrame* pFrame = RwFrameCreate();
+    RpAtomic* pClonedAtomic = RpAtomicClone(firstAtomic);
+    RpAtomicSetFrame(pClonedAtomic, pFrame);
+
+    m_pInterface->pRwObject = nullptr;
+    ((void(__thiscall*)(CAtomicModelInfoSAInterface * pThis, RpAtomic * pAtomic)) vfbl->SetAtomic)(m_pInterface, pClonedAtomic);
+
+    ppModelInfo[m_dwModelID] = m_pInterface;
+
+    RpClumpDestroy(pClump);
+    m_dwParentID = -1;
+    return true;
+}
+
+bool CModelInfoSA::MakeClumpModel(ushort usModelId)
+{
+    if (GetModelType() != eModelInfoType::ATOMIC)
+        return false;
+
+    CBaseModelInfoSAInterface*  pBaseObjectInfo = (CBaseModelInfoSAInterface*)ppModelInfo[usModelId];
+    CClumpModelInfoSAInterface* m_pInterface = new CClumpModelInfoSAInterface();
+    MemCpyFast(m_pInterface, pBaseObjectInfo, sizeof(CBaseModelInfoSAInterface));
+    m_pInterface->m_nAnimFileIndex = -1;
+
+    CClumpModelInfo_SA_VTBL* vfbl = (CClumpModelInfo_SA_VTBL*)vftable_CClumpModelModelInfo;
+    m_pInterface->VFTBL = vfbl;
+    RpClump*  pClump = RpClumpCreate();
+    RpAtomic* pAtomic = (RpAtomic*)m_pInterface->pRwObject;
+    RpSetFrame(pClump, RwFrameCreate());
+    RpClumpAddAtomic(pClump, pAtomic);
+
+    ((void(__thiscall*)(CClumpModelInfoSAInterface * pThis, RpClump * clump)) vfbl->SetClump)(m_pInterface, pClump);
+
+    ppModelInfo[m_dwModelID] = m_pInterface;
```

Comment:
I see a potential memory leak in here. Because you're creating a new instance of CClumpModelInfoSAInterface, but the previous one keep living in memory.